lightly/loss/ibot_loss.py

```diff
@@ -0,0 +1,85 @@
+import torch
+from torch import Tensor
+from torch.nn import Module
+from torch.nn import functional as F
+
+from lightly.models.modules.center import Center
+
+
+class IBOTPatchLoss(Module):
+    """Implementation of the iBOT patch loss [0] as used in DINOv2 [1].
+
+    Implementation is based on [2].
+
+    - [0]: iBOT, 2021, https://arxiv.org/abs/2111.07832
+    - [1]: DINOv2, 2023, https://arxiv.org/abs/2304.07193
+    - [2]: https://github.com/facebookresearch/dinov2/blob/main/dinov2/loss/ibot_patch_loss.py
+
+    Attributes:
+        temperature:
+            Temperature of the softmax for the student outputs.
+    """
+
+    def __init__(
+        self,
+        output_dim: int,
+        teacher_temp: float = 0.04,
+        student_temp: float = 0.1,
+        center_mode: str = "mean",
+        center_momentum: float = 0.9,
+    ) -> None:
+        super().__init__()
+        self.teacher_temp = teacher_temp
+        self.student_temperature = student_temp
+        self.center = Center(
+            size=(1, output_dim),
+            mode=center_mode,
+            momentum=center_momentum,
+        )
+
+    def forward(
+        self,
+        teacher_out: Tensor,
+        student_out: Tensor,
+        mask: Tensor,
+    ) -> Tensor:
+        """Forward pass through the iBOT patch loss.
+
+        Args:
+            teacher_out:
+                Tensor with shape (B * N, D) containing the teacher output of the masked
+                tokens.
+            student_out:
+                Tensor with shape (B * N, D) containing the student output of the masked
+                tokens.
+            mask:
+                Boolean tensor with shape (B, H, W) containing the token mask.
+                Exactly B * N entries must be set to True in the mask.
```

Comment:
The DINOv2 implementation doesn't use the `forward` method of the loss, it only uses the `forward_masked` method which has the following shapes:
```
        student_patch_tokens_masked: (B * N , D)
        teacher_patch_tokens_masked: (B * N, D)
        student_masks_flat: (B, H * W)
```

The only difference is the shape of the `student_masks_flat` shape. I kept it as `(B, H, W)` because this is how the mask will be shaped everywhere in the code.